### PR TITLE
Code: divide charge_mixing.cpp into some files

### LINKED ISSUE #5425: Code: refactor `charge_mixing.cpp`
### Describe the Code Quality Issue

This file is too long

### Additional Context

_No response_

### Task list for Issue attackers (only for developers)

- [ ] Identify the specific code file or section with the code quality issue.
- [ ] Investigate the issue and determine the root cause.
- [ ] Research best practices and potential solutions for the identified issue.
- [ ] Refactor the code to improve code quality, following the suggested solution.
- [ ] Ensure the refactored code adheres to the project's coding standards.
- [ ] Test the refactored code to ensure it functions as expected.
- [ ] Update any relevant documentation, if necessary.
- [ ] Submit a pull request with the refactored code and a description of the changes made.

### PR DESCRIPTION
**Pull request by WHUweiqingzhou**
Fix #5425

### CODE REVIEW

File: source/module_elecstate/module_charge/charge_mixing_dmr.cpp [hunk 1]
```diff
@@ -0,0 +1,227 @@
+#include "charge_mixing.h"
+
+#include "module_parameter/parameter.h"
+#include "module_base/timer.h"
+
+void Charge_Mixing::allocate_mixing_dmr(int nnr)
```
Comment (mohanchen) [inline source/module_elecstate/module_charge/charge_mixing_dmr.cpp]: add const before "int nnr"?

File: source/module_elecstate/module_charge/charge_mixing_dmr.cpp [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+{
+    // Notice that DensityMatrix object is a Template class
+    ModuleBase::TITLE("Charge_Mixing", "mix_dmr");
+    ModuleBase::timer::tick("Charge_Mixing", "mix_dmr");
+    //
+    std::vector<hamilt::HContainer<double>*> dmr = DM->get_DMR_vector();
+    std::vector<std::vector<double>>& dmr_save = DM->get_DMR_save();
+    //
+    //const int dmr_nspin = (PARAM.inp.nspin == 2) ? 2 : 1;
+    double* dmr_in;
```
Comment (mohanchen) [inline source/module_elecstate/module_charge/charge_mixing_dmr.cpp]: initialize with null?

File: source/module_elecstate/module_charge/charge_mixing_dmr.cpp [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
```
Comment (mohanchen) [inline source/module_elecstate/module_charge/charge_mixing_dmr.cpp]: initialize with null?

File: source/module_elecstate/module_charge/charge_mixing_uspp.cpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    ModuleBase::TITLE("Charge_Mixing", "clean_data");
+    if (PARAM.inp.nspin == 1)
+    {
+        data_s = nullptr;
+        data_hf = nullptr;
+        return;
+    }
+    else
+    {
+        delete[] data_s;
```
Comment (mohanchen) [inline source/module_elecstate/module_charge/charge_mixing_uspp.cpp:71]: seems dangerous to delete pointer, can we use std::vector instead?

File: source/module_elecstate/module_charge/charge_mixing_uspp.cpp [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    {
+        const int ndimd = this->rhodpw->npw;
+        const int ndims = this->rhopw->npw;
+        const int ndimhf = ndimd - ndims;
+        for (int is = 0; is < PARAM.inp.nspin; ++is)
+        {
+            std::memcpy(data_d + is * ndimd, data_s + is * ndims, ndims * sizeof(std::complex<double>));
+            std::memcpy(data_d + is * ndimd + ndims, data_hf + is * ndimhf, ndimhf * sizeof(std::complex<double>));
+        }
+        delete[] data_s;
```
Comment (mohanchen) [inline source/module_elecstate/module_charge/charge_mixing_uspp.cpp:53]: seems dangerous to delete pointer, can we use std::vector instead?

File: source/module_elecstate/module_charge/charge_mixing_uspp.cpp (hunk 4 above)
Comment (WHUweiqingzhou) [reply source/module_elecstate/module_charge/charge_mixing_uspp.cpp:71]: > seems dangerous to delete pointer, can we use std::vector instead?

@YuLiu98 what's your idea?

File: source/module_elecstate/module_charge/charge_mixing_preconditioner.cpp [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+#include "module_parameter/parameter.h"
+#include "module_base/timer.h"
+#include "module_hamilt_pw/hamilt_pwdft/global.h"
+
+void Charge_Mixing::Kerker_screen_recip(std::complex<double>* drhog)
+{
+    if (this->mixing_gg0 <= 0.0 || this->mixing_beta <= 0.1) {
+        return;
+}
+    double fac, gg0, amin;
```
Comment (mohanchen) [inline source/module_elecstate/module_charge/charge_mixing_preconditioner.cpp]: initialize the variables, it is better to define one variable in one line

File: source/module_elecstate/module_charge/charge_mixing_preconditioner.cpp [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    }
+    return;
+}
+
+void Charge_Mixing::Kerker_screen_real(double* drhor)
+{
+    if (this->mixing_gg0 <= 0.0001 || this->mixing_beta <= 0.1) {
+        return;
+}
+    // consider a resize for mixing_angle
```
Comment (mohanchen) [inline source/module_elecstate/module_charge/charge_mixing_preconditioner.cpp]: I would suggest using doxygen-style notes

File: source/module_elecstate/module_charge/charge_mixing_preconditioner.cpp [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    std::vector<std::complex<double>> drhog(this->rhopw->npw * PARAM.inp.nspin / resize_tmp);
+    std::vector<double> drhor_filter(this->rhopw->nrxx * PARAM.inp.nspin / resize_tmp);
+    for (int is = 0; is < PARAM.inp.nspin / resize_tmp; ++is)
+    {
+        // Note after this process some G which is higher than Gmax will be filtered.
+        // Thus we cannot use Kerker_screen_recip(drhog.data()) directly after it.
+        this->rhopw->real2recip(drhor + is * this->rhopw->nrxx, drhog.data() + is * this->rhopw->npw);
+    }
+    // implement Kerker for density and magnetization separately
+    double fac, gg0, amin;
```
Comment (mohanchen) [inline source/module_elecstate/module_charge/charge_mixing_preconditioner.cpp]: initialize the variables, one variable for one line

File: source/module_elecstate/module_charge/charge_mixing_uspp.cpp (hunk 4 above)
Comment (YuLiu98) [reply source/module_elecstate/module_charge/charge_mixing_uspp.cpp:71]: I suggest to modify it by the way.

**Review by mohanchen — APPROVED** [review]

**Comment by mohanchen** [comment]
Some small places need to be updated, but this can be left for near future.

**Pull request merged**